plugins/catalog-backend/src/next/database/DefaultProcessingDatabase.ts

```diff
@@ -332,143 +274,40 @@ export class DefaultProcessingDatabase implements ProcessingDatabase {
-        sourceKey: options.sourceKey,
-        entities: toAdd,
-      });
-    }
-  }
+      for (const { entity, locationKey } of toAdd) {
+        const entityRef = stringifyEntityRef(entity);
 
-  /**
-   * Add a set of deferred entities for processing.
-   * The entities will be added at the front of the processing queue.
-   */
-  async addUnprocessedEntities(
-    txOpaque: Transaction,
-    options: AddUnprocessedEntitiesOptions,
-  ): Promise<void> {
-    const tx = txOpaque as Knex.Transaction;
-
-    // Keeps track of the entities that we end up inserting to update refresh_state_references afterwards
-    const stateReferences = new Array<string>();
-    const conflictingStateReferences = new Array<string>();
-
-    // Upsert all of the unprocessed entities into the refresh_state table, by
-    // their entity ref.
-    for (const { entity, locationKey } of options.entities) {
-      const entityRef = stringifyEntityRef(entity);
-      const serializedEntity = JSON.stringify(entity);
-
-      // We optimistically try to update any existing refresh state first, as this is by far
-      // the most common case.
-      const refreshResult = await tx<DbRefreshStateRow>('refresh_state')
-        .update({
-          unprocessed_entity: serializedEntity,
-          location_key: locationKey,
-          last_discovery_at: tx.fn.now(),
-          // We only get to this point if a processed entity actually had any changes, or
-          // if an entity provider requested this mutation, meaning that we can safely
-          // bump the deferred entities to the front of the queue for immediate processing.
-          next_update_at: tx.fn.now(),
-        })
-        .where('entity_ref', entityRef)
-        .andWhere(inner => {
-          if (!locationKey) {
-            return inner.whereNull('location_key');
-          }
-          return inner
-            .where('location_key', locationKey)
-            .orWhereNull('location_key');
-        });
-
-      if (refreshResult === 0) {
-        // In the event that we can't update an existing refresh state, we first try to insert a new row
         try {
-          let query = tx('refresh_state').insert<any>({
-            entity_id: uuid(),
-            entity_ref: entityRef,
-            unprocessed_entity: serializedEntity,
-            errors: '',
-            location_key: locationKey,
-            next_update_at: tx.fn.now(),
-            last_discovery_at: tx.fn.now(),
-          });
-
-          // TODO(Rugvip): only tested towards Postgres and SQLite
-          // We have to do this because the only way to detect if there was a conflict with
-          // SQLite is to catch the error, while Postgres needs to ignore the conflict to not
-          // break the ongoing transaction.
-          if (tx.client.config.client !== 'sqlite3') {
-            query = query.onConflict('entity_ref').ignore();
+          let ok = await this.insertUnprocessedEntity(tx, entity, locationKey);
```

Comment:
Not to be fixed in this PR, but it is probably unfortunate that we try insert first and update after. Large refreshes of mostly pre-existing things will have a large amount of data in toAdd, and this code kinda assumes that add means add, not upsert.

It could be that we want to add a toUpdate array in the delta output. It could even check content hashes if we wanted to.

Reply:
I flipped them around here for the opposite reason. We compute the delta first so the by far most common case is that it doesn't already exist

Reply:
So when I issue a 'full' sync, what happens to all the ones that exist and have changes?

Reply:
Nothing, right now the full sync is only concerned with making sure the entities are(n't) in the catalog, it doesn't trigger refreshes other than by "accident" when an entity is already provided from another source